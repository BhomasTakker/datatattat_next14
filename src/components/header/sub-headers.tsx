import React, { ReactElement } from "react";
import { NavigationMenu } from "./navigation/navigation-menu";
import { HeaderType } from "@/types/header";

const renderSubMenu = (item: HeaderType, index: number): ReactElement => {
	const { nav = [] } = item || {};
	return <NavigationMenu items={nav} key={index} />;
};

const renderList = (list: HeaderType[]): ReactElement => {
	return <div>{list.map(renderSubMenu)}</div>;
};

const trimList = (list: HeaderType[]) => {
	const { length } = list;
	// Should use a config value?
	return length <= 2 ? list : list.slice(length - 2);
};

export const SubHeaders = ({
	headersArray,
}: {
	headersArray: HeaderType[];
}) => {
	if (!headersArray || headersArray.length === 0) {
		return null;
	}

	// Take in list and reverse
	const reversedList = headersArray.reverse();

	const trimmedList = trimList(reversedList);

	return renderList(trimmedList);
};
