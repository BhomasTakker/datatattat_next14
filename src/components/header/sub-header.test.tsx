import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";

import { SubHeaders } from "./sub-headers";

type Props = { items: { label: string; route: string }[]; key: number };

jest.mock("./navigation/navigation-menu", () => {
	return {
		NavigationMenu: ({ items }: Props) => {
			return (
				<div>
					<ul>
						{items.map((item) => (
							<li key={item.label}>
								<p>{item.label}</p>
								<p>{item.route}</p>
							</li>
						))}
					</ul>
				</div>
			);
		},
	};
});

const mock = (id: string | number) => ({
	route: `/test-route-${id}`,
	nav: [
		{
			label: "Home",
			route: "/",
		},
		{
			label: "About",
			route: "/about",
		},
	],
});

describe("Sub Headers Test Suite", () => {
	it("renders a header", () => {
		render(SubHeaders({ headersArray: [mock("mock")] }));
		const home = screen.getByText("Home");
		expect(home).toBeInTheDocument();
	});

	it("renders 2 sub headers", () => {
		render(SubHeaders({ headersArray: [mock("mock1"), mock("mock2")] }));
		const homes = screen.getAllByText("Home");
		expect(homes.length).toEqual(2);
	});

	it("renders no more than 2 sub headers", () => {
		render(
			SubHeaders({
				headersArray: [
					mock("mock1"),
					mock("mock2"),
					mock("mock3"),
					mock("mock4"),
				],
			})
		);
		const homes = screen.getAllByText("Home");
		expect(homes.length).toEqual(2);
	});

	it("does not render if given headers array is empty", () => {
		const { container } = render(<SubHeaders headersArray={[]} />);
		expect(container).toBeEmptyDOMElement();
	});

	// How do we tst the order of the sub headers?

	it("renders SubHeaders unchanged", async () => {
		const { container } = render(
			<SubHeaders headersArray={[mock("mock1"), mock("mock2")]} />
		);
		expect(container).toMatchSnapshot();
	});
});
