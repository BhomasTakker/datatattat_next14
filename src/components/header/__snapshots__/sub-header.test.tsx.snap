// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Sub Headers Test Suite renders SubHeaders unchanged 1`] = `
<div>
  <div>
    <div>
      <ul>
        <li>
          <p>
            Home
          </p>
          <p>
            /
          </p>
        </li>
        <li>
          <p>
            About
          </p>
          <p>
            /about
          </p>
        </li>
      </ul>
    </div>
    <div>
      <ul>
        <li>
          <p>
            Home
          </p>
          <p>
            /
          </p>
        </li>
        <li>
          <p>
            About
          </p>
          <p>
            /about
          </p>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
